snekbox/snekio.py

```diff
@@ -0,0 +1,101 @@
+"""I/O Operations for sending / receiving files from the sandbox."""
+from __future__ import annotations
+
+from base64 import b64decode, b64encode
+from dataclasses import dataclass
+from functools import cached_property
+from pathlib import Path
+from typing import Generic, TypeVar
+
+T = TypeVar("T", str, bytes)
+
+
+def safe_path(path: str) -> str:
+    """
+    Returns the `path` str if there are no security issues.
+
+    Raises:
+        IllegalPathError: Raised on any path rule violation.
+    """
+    # Disallow absolute paths
+    if Path(path).is_absolute():
+        raise IllegalPathError(f"File path '{path}' must be relative")
+
+    # Disallow traversal beyond root
+    try:
+        test_root = Path("/home")
+        Path(test_root).joinpath(path).resolve().relative_to(test_root.resolve())
+    except ValueError:
+        raise IllegalPathError(f"File path '{path}' may not traverse beyond root")
+
+    return path
+
+
+class AttachmentError(ValueError):
+    """Raised when an attachment is invalid."""
+
+
+class ParsingError(AttachmentError):
+    """Raised when an incoming file cannot be parsed."""
+
+
+class IllegalPathError(ParsingError):
+    """Raised when an attachment has an illegal path."""
+
+
+@dataclass
+class FileAttachment(Generic[T]):
+    """A file attachment."""
+
+    path: str
+    content: T
+
+    @classmethod
+    def from_dict(cls, data: dict[str, str]) -> FileAttachment[bytes]:
+        """Convert a dict to an attachment."""
+        path = safe_path(data["path"])
```

Comment:
I think the only invalid path is a single `/` on linux? And that would be caught by as an absolute path. File writing is also in a `try` block with an associated `OSError` report.

Though I have added handling for absolute paths (any path beginning with `/` or `\0`) and unit tests for these in [7e7c0d6](https://github.com/python-discord/snekbox/pull/159/commits/7e7c0d6f5465f965b3ee31fc4f723ec23cc5ad94)

I am not sure if we can do traversal checks with regex. And I think we should still keep the current pathlib checks for absolute and traversal (since the json schema validator is not used for `__main__` calls and library use)

Reply:
Added additional unit tests in [dffd9c8](https://github.com/python-discord/snekbox/pull/159/commits/dffd9c8100230d6e8df8c0261fbd5ae8dbd0ceb2)

Reply:
Updated regex to
```regex
^(?!/)(?!.*\\0).*$
```
This should disallow anything beginning with `/` or `\0` contained anywhere.